src/cljam/io/bam/writer.clj

```diff
@@ -53,14 +78,46 @@
       (lsb/write-int w (:len ref)))))
 
 (defn write-alignments* [^BAMWriter wtr alns header]
-  (let [w (.writer wtr)
+  (let [dw (.data-writer wtr)
+        w ^BGZFOutputStream (.writer wtr)
         refs (refs/make-refs header)]
-    (doseq [a alns]
-      (lsb/write-int w (encoder/get-block-size a))
-      (encoder/encode-alignment w a refs))))
+    (with-open [baos (java.io.ByteArrayOutputStream. 4096)
+                dos (java.io.DataOutputStream. baos)]
+      (let [pointer-block (map
+                           (fn [a]
+                             (let [pointer-beg (.getFilePointer w)]
+                               (.reset baos)
+                               (encoder/encode-alignment dos a refs)
+                               (lsb/write-int dw (.size baos))
+                               (.writeTo baos dw)
+                               (when @(.index wtr)
+                                 (bam-decoder/decode-pointer-block
+                                  (BAMRawBlock. (.toByteArray baos)
+                                                pointer-beg
+                                                (.getFilePointer w))))))
+                           alns)]
+        (if @(.index wtr)
+          (swap! (.index wtr) bai-writer/merge-index
```

Comment:
I'm so sorry 🤦  We can't use `swap!` for this case as it retains the argument while trying CAS. Would you revert the change, please?

Reply:
Fixed.